drivers/platform/aducm3029/flash.c

```diff
@@ -0,0 +1,289 @@
+ * @param [in] data       - Pointer to the data to be written.
+ *                          data length must be FLASH_PAGE_SIZE_WORDS in size.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t flash_write_page(struct flash_dev *dev, int32_t page_no, uint32_t *data)
+{
+	ADI_FEE_TRANSACTION transaction;
+	struct adicup_flash_dev *adicup_extra = dev->extra;
+	uint32_t fee_hw_error;
+	int32_t ret;
+
+	transaction.bUseDma = true;
+	transaction.nSize = FLASH_PAGE_SIZE_BYTES;
+	transaction.pWriteAddr = (uint32_t *)(page_no << FLASH_PAGE_ADDR_SHIFT);
+	transaction.pWriteData = data;
+
+	ret = adi_fee_Write(adicup_extra->instance, &transaction,
+			    &fee_hw_error);
+	if(ret != ADI_FEE_SUCCESS)
+		return FAILURE;
+
+	return SUCCESS;
+}
+
+/**
+ * Read-modify-write helper function for flash_write(). Since the smallest write
+ * unit for the ADuCM3029 is the page this function helps access data smaller than
+ * a page or accross multiple pages.
+ */
+static int32_t flash_read_then_write(struct flash_dev *dev, uint32_t flash_addr,
+				     uint32_t *array, uint32_t array_size)
+{
+	struct adicup_flash_dev *adicup_extra = dev->extra;
+	uint32_t page_nr, fee_hw_error;
+	uint32_t temp_ptr[FLASH_PAGE_SIZE_WORDS] __attribute__ ((aligned (sizeof(
+				uint32_t))));
+	int32_t ret;
+
+	if ((flash_addr & 0x3) != 0)
+		return FAILURE;
+	if ((FLASH_OFFSET_IN_PAGE(flash_addr) + array_size) >
+	    FLASH_PAGE_SIZE_WORDS)
+		return FAILURE;
+
+	flash_read(dev, FLASH_ADDRESS_PAGE_START(flash_addr), temp_ptr,
+		   FLASH_PAGE_SIZE_WORDS);
+
+	memcpy(temp_ptr + FLASH_OFFSET_IN_PAGE(flash_addr), array,
+	       array_size * sizeof(uint32_t));
+
+	/* Get the page number */
+	ret = adi_fee_GetPageNumber(adicup_extra->instance, flash_addr,
+				    &page_nr);
+	if(ret != ADI_FEE_SUCCESS)
+		return FAILURE;
+
+	/* First erase page */
+	ret = adi_fee_PageErase(adicup_extra->instance, page_nr, page_nr,
+				&fee_hw_error);
+	if(ret != ADI_FEE_SUCCESS)
+		return FAILURE;
+
+	return flash_write_page(dev, page_nr, temp_ptr);
+}
+
+/**
+ * Write data in flash memory.
+ *
+ * @param [in] dev        - Pointer to the flash device handler.
+ * @param [in] flash_addr - Start address of the write.
+ * @param [in] array      - Pointer to the data to be written.
+ * @param [in] array_size - Size of the written data in 32-bit words.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t flash_write(struct flash_dev *dev, uint32_t flash_addr,
+		    uint32_t *array, uint32_t array_size)
+{
+	uint32_t data_index = 0, temp, initial_size = array_size;
+	int32_t ret;
+
+	do {
+		temp = FLASH_PAGE_SIZE_WORDS - FLASH_OFFSET_IN_PAGE(flash_addr);
+		if (temp > array_size)
+			temp = array_size;
+		ret = flash_read_then_write(dev, flash_addr, (array + data_index),
+					    temp);
+		if (ret != SUCCESS)
+			return FAILURE;
+		flash_addr += temp;
+		array_size -= temp;
+		data_index += temp;
+	} while (data_index < initial_size);
```

Comment:
nitpcik: I would simplify this to don't have these many variables if it makes sense
```suggestion
        i = 0;
	while (i < array_size) {
		len = FLASH_PAGE_SIZE_WORDS - FLASH_OFFSET_IN_PAGE(flash_addr + i);
		if (len > array_size - i)
			len = array_size - i;
		ret = flash_read_then_write(dev, flash_addr + i, array + i, len);
		if (ret != SUCCESS)
			return FAILURE;
		i += len;
	}
```